### PR TITLE
[SR GeekCamp]support admin set replica status syntax

### LINKED ISSUE #4043: New Parser and Analyzer support ADMIN SET related statements【Easy】
# Requirement Description
The new Parser and the new Analyzer support admin set operation related statements

# Mentor
@HangyuanLiu 

# Registration procedure

**Step 1.** Add a comment when you sign up. The comment can be a brief self-introduction and reason for signing up, including what you do, what you're good at, what you're interested in and your technology stack, and how familiar you are with StarRocks. Example:

> I pick! My name is Johnny. I am from StarRocks and I have been working on database kernel development for 1 year. I am interested in OLAP, good at C++ programming, and have been using StarRocks for 6 months. I chose to solve this issue because I am familiar with execution engine.

**Step 2.** On your WeChat, add the StarRocks community assistant "**_StarRocks-1_**" as a WeChat friend and leave your GitHub ID. The assistant will invite you to a WeChat group based on the issue you picked.

**Step 3.** The mentor will communicate with all the participants who picked this issue and assign the task to the best match. 

# Before you start
1. Read the StarRocks [GitHub workflow](https://github.com/StarRocks/community/blob/main/Contributors/guide/workflow.md) carefully, which provides the configurations of BE and FE and describes how to submit a pull request.
2. To learn more about StarRocks, read [StarRocks documentation](https://github.com/StarRocks/docs/blob/main/TOC.md).

# Pull request requirements

1. After you complete code development, upload the code to the StarRocks code repo as a pull request (PR).

  - PR title format: **[SR GeekCamp] + PR title**
  - Labels ：[Geek Camp 2022/feature]
2. After you submit your PR, comment on the status of your PR submission as requested under the Pinned Issue [[StarRocks geek Camp 2022] ](https://github.com/StarRocks/starrocks/issues/4008)page.

  - Comment format: **"I have finished Issue #" + PR link id**

### PR DESCRIPTION
**Pull request by dreamay**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #4043 

## Problem Summary(Required) ：

- Support admin set replica status syntax
- Replace cup with antlr4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 17 / 21 (80.95%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AdminSetReplicaStatusStmtAnalyzer.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AdminSetReplicaStatusStmt.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 18 / 23 (78.26%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AdminSetReplicaStatusStmtAnalyzer.java|9|13|69.23%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AdminSetReplicaStatusStmt.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 17 / 21 (80.95%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AdminSetReplicaStatusStmtAnalyzer.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AdminSetReplicaStatusStmt.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 19 / 24 (79.17%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AdminSetReplicaStatusStmtAnalyzer.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AdminSetReplicaStatusStmt.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 21 / 27 (77.78%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|4|6|66.67%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AdminSetReplicaStatusStmtAnalyzer.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AdminSetReplicaStatusStmt.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 21 / 27 (77.78%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|4|6|66.67%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AdminSetReplicaStatusStmtAnalyzer.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AdminSetReplicaStatusStmt.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 21 / 27 (77.78%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|4|6|66.67%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AdminSetReplicaStatusStmtAnalyzer.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AdminSetReplicaStatusStmt.java|1|1|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 1]
```diff
@@ -34,10 +34,11 @@ statement
         comment? AS queryStatement                                                          #createView
     | ALTER VIEW qualifiedName
         ('(' columnNameWithComment (',' columnNameWithComment)* ')')?
-        AS queryStatement                                                                   #alterView
-    | DROP TABLE (IF EXISTS)? qualifiedName FORCE?                                          #dropTable
-    | DROP VIEW (IF EXISTS)? qualifiedName                                                  #dropView
-    | ADMIN SET FRONTEND CONFIG '(' property ')'                                            #adminSetConfig
+        AS queryStatement                                                               #alterView
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:37]: Note the indented format

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AdminSetReplicaStatusStmtAnalyzer.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    static class AdminSetStmtAnalyzerVisitor extends AstVisitor<Void, ConnectContext> {
+        public void analyze(AdminSetReplicaStatusStmt adminSetReplicaStatusStmt, ConnectContext session) {
+            visit(adminSetReplicaStatusStmt, session);
+        }
+
+        @Override
+        public Void visitAdminSetReplicaStatusStatement(AdminSetReplicaStatusStmt adminSetReplicaStatusStmt,
+                                                        ConnectContext session) {
+            try {
+                adminSetReplicaStatusStmt.checkProperties();
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AdminSetReplicaStatusStmtAnalyzer.java:23]: The logic of checkProperties needs to be extracted, the new version does not depend on the code in the old version

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 43 / 54 (79.63%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|4|6|66.67%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AdminSetReplicaStatusStmtAnalyzer.java|29|37|78.38%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AdminSetReplicaStatusStmt.java|2|2|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AdminSetReplicaStatusStmtAnalyzer.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        public void analyze(AdminSetReplicaStatusStmt adminSetReplicaStatusStmt, ConnectContext session) {
+            visit(adminSetReplicaStatusStmt, session);
+        }
+
+        @Override
+        public Void visitAdminSetReplicaStatusStatement(AdminSetReplicaStatusStmt adminSetReplicaStatusStmt,
+                                                        ConnectContext session) {
+            try {
+                checkProperties(adminSetReplicaStatusStmt);
+            } catch (AnalysisException e) {
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AdminSetReplicaStatusStmtAnalyzer.java:27]: No need to catch AnalysisException, directly change to throw SemanticException

File: fe/fe-core/src/test/java/com/starrocks/analysis/AdminSetReplicaStatusStmtTest.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import com.starrocks.catalog.Catalog;
+import com.starrocks.common.AnalysisException;
+import com.starrocks.qe.ConnectContext;
+import com.starrocks.utframe.UtFrameUtils;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+public class AdminSetReplicaStatusStmtTest {
```
Comment (HangyuanLiu) [inline fe/fe-core/src/test/java/com/starrocks/analysis/AdminSetReplicaStatusStmtTest.java:12]: Merge UT into AdminSetTest.java

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 43 / 54 (79.63%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|4|6|66.67%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AdminSetReplicaStatusStmtAnalyzer.java|29|37|78.38%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AdminSetReplicaStatusStmt.java|2|2|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AdminSetReplicaStatusStmtAnalyzer.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        private void checkProperties(AdminSetReplicaStatusStmt adminSetReplicaStatusStmt) {
+            long tabletId = -1;
+            long backendId = -1;
+            ReplicaStatus status = null;
+            Map<String, String> properties = adminSetReplicaStatusStmt.getProperties();
+            for (Map.Entry<String, String> entry : properties.entrySet()) {
+                String key = entry.getKey();
+                String val = entry.getValue();
+
+                if (key.equalsIgnoreCase("tablet_id")) {
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AdminSetReplicaStatusStmtAnalyzer.java:41]:
```suggestion
                if (key.equalsIgnoreCase(AdminSetReplicaStatusStmt.TABLET_ID)) {
```

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 43 / 54 (79.63%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|4|6|66.67%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AdminSetReplicaStatusStmtAnalyzer.java|29|37|78.38%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AdminSetReplicaStatusStmt.java|2|2|100.00%|

**Review by HangyuanLiu — DISMISSED** [review]

**Review by HangyuanLiu — DISMISSED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Review by liuyehcf — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Pull request merged**